File: src/simulation/projectile.ts
import type { Island } from '../maps/catalog';
import { firstLandHit } from './land';
import type { Vec3 } from '../ships/blueprint';
import type { FleetActor } from './battle';
import { ballisticStep, travelFactor, velocityPenetration } from './ballistics';
import { burstShell } from './burst';
import { resolveShipContact, shipContacts, type DamageEvent, type Shell, type ShipContact } from './damage';
import { add, length, localToWorld, radians, scale, segmentOverlapsBox, sub, worldToLocal } from './geometry';
import { hullContains } from './hull';
import { mayReachHull, shellHullRadius } from './spatial';

export type ProjectileEnd = 'burst' | 'stopped' | 'passed-through' | 'splash' | 'expired';
/** Swept chords remain bounded to a CPU tick. Every contact splits elapsed time,
 * so residual speed and fuze delay affect travel during that same tick. */
export function advanceProjectile(shell: Shell, actors: FleetActor[], dt: number, emit: (event: DamageEvent | { kind: 'splash'; shipId: string; position: Vec3; message: string; shell: Pick<Shell, 'id' | 'caliberM' | 'velocity' | 'type' | 'ammunition'> }) => void, islands: readonly Island[] = [], surfaceAt: (x: number, z: number) => number = () => 0): ProjectileEnd | undefined {
  const insideHull = (point: Vec3) => actors.some(a => a.motion.id !== shell.ownerId && hullContains(a.definition.hull, worldToLocal(point, a.motion)));
  let remaining = dt;
  for (let iteration = 0; iteration < 64; iteration++) {
    if (shell.lodged) {
      const actor = actors.find(a => a.motion.id === shell.lodged!.shipId);
      if (actor) {
        const index = actor.definition.mounts.findIndex(m => m.id === shell.lodged!.mountId), mount = actor.definition.mounts[index];
        const local = mount ? localToWorld(shell.lodged.position, { x: mount.position[0], y: mount.position[1], z: mount.position[2], heading: radians(mount.bearingDeg) + actor.mounts[index].train, roll: 0, pitch: 0 }) : shell.lodged.position;
        shell.position = localToWorld(local, actor.motion);
      }
    }
    if (shell.detonateAtAge !== undefined && shell.age >= shell.detonateAtAge - 1e-10) {
      burstShell(shell, actors, emit); return 'burst';
    }
    if (remaining <= 1e-10) return;
    if (shell.age >= 180) return 'expired';
    let horizon = Math.min(remaining, 180 - shell.age, shell.detonateAtAge === undefined ? remaining : shell.detonateAtAge - shell.age);
    if (shell.lodged) { shell.age += horizon; remaining -= horizon; continue; }
    const from: Vec3 = [...shell.position];
    const inWater = from[1] <= surfaceAt(from[0], from[2]) && !insideHull(from);
    if (inWater && length(shell.velocity) < 20 && shell.detonateAtAge === undefined) return shell.visited.length ? 'passed-through' : 'splash';
    if (inWater) shell.waterDragPerSecond ??= Math.max(1, length(shell.velocity) * .04 * .38 / Math.max(.007, shell.caliberM));
    const drag = inWater ? shell.waterDragPerSecond! : shell.dragPerSecond ?? 0;
    let flight = ballisticStep(from, shell.velocity, horizon, drag);
    // Split at the exact sea crossing; never advance the fuze through the air
    // portion using an underwater velocity, or discard the remaining tick.
    let crossesSea = false;
    const endsInWater = flight.position[1] <= surfaceAt(flight.position[0], flight.position[2]);
    if (inWater !== endsInWater && (inWater || from[1] > surfaceAt(from[0], from[2]))) {
      let low = 0, high = horizon;
      for (let i = 0; i < 40; i++) {
        const mid = (low + high) / 2;
        const point = ballisticStep(from, shell.velocity, mid, drag).position;
        if ((point[1] <= surfaceAt(point[0], point[2])) === inWater) low = mid; else high = mid;
      }
      const crossing = ballisticStep(from, shell.velocity, high, drag);
      if (!insideHull(crossing.position)) { horizon = high; flight = crossing; crossesSea = true; }
    }
    const end = flight.position;
    let nearest: { actor: FleetActor; hit: ShipContact } | undefined;
    for (const actor of actors) {
      if (actor.motion.id === shell.ownerId || !mayReachHull(from, end, actor.motion, shellHullRadius(actor.definition))) continue;
      const def = actor.definition;
      if (!segmentOverlapsBox(worldToLocal(from, actor.motion), worldToLocal(end, actor.motion), { center: [0, 10, 0], size: [def.hull.beam + 30, 60, def.hull.length + 40] })) continue;
      const hit = shipContacts(shell, from, end, actor, def)[0];
      if (hit && (!nearest || hit.t < nearest.hit.t)) nearest = { actor, hit };
    }
    const land = firstLandHit(islands, from, end);
    if (land && (!nearest || land.t < nearest.hit.t)) {
      shell.position = land.point;
      emit({ kind: 'stopped', shipId: '', position: land.point, message: 'Shell struck the coast', normal: [0, 1, 0], shell: { id: shell.id, caliberM: shell.caliberM, velocity: [...shell.velocity], ammunition: shell.ammunition, type: shell.type ?? 'AP' } });
      return 'stopped';
    }
    if (nearest) {
      const t = nearest.hit.t;
      const elapsed = drag > 1e-9 ? -Math.log1p(-drag * t * travelFactor(horizon, drag)) / drag : horizon * t;
      const atHit = ballisticStep(from, shell.velocity, elapsed, drag);
      shell.penetrationMm = velocityPenetration(shell.penetrationMm, length(shell.velocity), length(atHit.velocity));
      shell.velocity = atHit.velocity; shell.age += elapsed; remaining -= elapsed;
      shell.position = localToWorld(nearest.hit.point, nearest.actor.motion);
      const stopped = resolveShipContact(shell, nearest.hit, nearest.actor, nearest.actor.definition, emit);
      if (stopped && !shell.lodged && shell.detonateAtAge === undefined) return 'stopped';
      continue;
    }
    shell.penetrationMm = velocityPenetration(shell.penetrationMm, length(shell.velocity), length(flight.velocity));
    shell.velocity = flight.velocity; shell.position = end; shell.age += horizon; remaining -= horizon;
    if (crossesSea && inWater) {
      shell.position[1] = surfaceAt(end[0], end[2]) + 1e-7;
      shell.waterDragPerSecond = undefined;
    } else if (crossesSea) {
      shell.position[1] = surfaceAt(end[0], end[2]) - 1e-7;
      emit({ kind: 'splash', position: [end[0], surfaceAt(end[0], end[2]), end[2]], shipId: '', message: 'Shell entering water', shell: { id: shell.id, caliberM: shell.caliberM, velocity: [...shell.velocity], ammunition: shell.ammunition, type: shell.type ?? (shell.ammunition === 'he' ? 'HE' : 'AP') } });
      if (shell.he) { shell.detonateAtAge = shell.age; continue; }
      // Water impact strips energy before the high-resistance underwater run.
      const before = length(shell.velocity);
      shell.velocity = scale(shell.velocity, .75);
      shell.penetrationMm = velocityPenetration(shell.penetrationMm, before, length(shell.velocity));
      shell.waterDragPerSecond = Math.max(1, length(shell.velocity) * .04 * .38 / Math.max(.007, shell.caliberM));
      if (shell.ap && shell.detonateAtAge === undefined && Math.abs(shell.velocity[1]) / length(shell.velocity) > .15)
        shell.detonateAtAge = shell.age + shell.ap.fuzeDelaySeconds;
    }

  }
  // Malformed/extremely dense definitions cannot monopolize a simulation tick.
  return 'expired';
}

File: crates/naval-sim/src/projectile.rs
use crate::{
    ballistics::{ballistic_step, travel_factor, velocity_penetration},
    burst::burst_shell,
    contacts::{ShipContact, ship_contacts},
    definition::Vec3,
    environment::{Island, TerrainField, first_land_hit},
    geometry::*,
    hull::hull_contains,
    impact::{DamageEvent, ShellEffect, resolve_ship_contact},
    machinery::equipment_pose,
    shell::Shell,
    vessel::Vessel,
};
#[derive(Clone, Copy, Debug, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum ProjectileEnd {
    Burst,
    Stopped,
    PassedThrough,
    Splash,
    Expired,
}
fn inside_hull(point: Vec3, actors: &[Vessel], owner: &str) -> bool {
    actors.iter().any(|a| {
        a.motion.id != owner
            && hull_contains(&a.definition().hull, world_to_local(point, a.motion.pose()))
    })
}
/// Every contact splits elapsed time. A single bounded tick can cross air, armor and water.
pub fn advance_projectile(
    shell: &mut Shell,
    actors: &mut [Vessel],
    dt: f64,
    islands: &[Island],
    fields: &[TerrainField],
    surface_at: &impl Fn(f64, f64) -> f64,
) -> (Option<ProjectileEnd>, Vec<DamageEvent>) {
    let mut remaining = dt;
    let mut events = vec![];
    for _ in 0..64 {
        if let Some(l) = &shell.lodged
            && let Some(actor) = actors.iter().find(|a| a.motion.id == l.ship_id)
        {
            let def = actor.definition();
            let mount = l
                .mount_id
                .as_ref()
                .and_then(|id| def.mounts.iter().position(|m| &m.id == id));
            let module = l
                .module_id
                .as_ref()
                .and_then(|id| def.modules.iter().find(|m| &m.id == id));
            let pose = module
                .and_then(|m| equipment_pose(actor, def, m))
                .or_else(|| {
                    mount.map(|i| {
                        let m = &def.mounts[i];
                        Pose {
                            x: m.position[0],
                            y: m.position[1],
                            z: m.position[2],
                            heading: radians(m.bearing_deg) + actor.mounts[i].train,
                            ..Pose::default()
                        }
                    })
                });
            let local = pose.map_or(l.position, |p| local_to_world(l.position, p));
            shell.position = local_to_world(local, actor.motion.pose());
        }
        if shell
            .detonate_at_age
            .is_some_and(|age| shell.age >= age - 1e-10)
        {
            let surface = surface_at(shell.position[0], shell.position[2]);
            let water = shell.position[1] <= surface
                && !inside_hull(shell.position, actors, &shell.owner_id);
            let mut burst = burst_shell(shell, actors);
            if water {
                for e in &mut burst {
                    e.water_burst_y = Some(surface);
                }
            }
            events.extend(burst);
            return (Some(ProjectileEnd::Burst), events);
        }
        if remaining <= 1e-10 {
            return (None, events);
        }
        if shell.age >= 180.0 {
            return (Some(ProjectileEnd::Expired), events);
        }
        let mut horizon = remaining.min(180.0 - shell.age).min(
            shell
                .detonate_at_age
                .map_or(remaining, |age| age - shell.age),
        );
        if shell.lodged.is_some() {
            shell.age += horizon;
            remaining -= horizon;
            continue;
        }
        let from = shell.position;
        let in_water =
            from[1] <= surface_at(from[0], from[2]) && !inside_hull(from, actors, &shell.owner_id);
        if in_water && length(shell.velocity) < 20.0 && shell.detonate_at_age.is_none() {
            return (
                Some(if shell.visited.is_empty() {
                    ProjectileEnd::Splash
                } else {
                    ProjectileEnd::PassedThrough
                }),
                events,
            );
        }
        if in_water {
            let drag =
                1.0_f64.max(length(shell.velocity) * 0.04 * 0.38 / shell.caliber_m.max(0.007));
            shell.water_drag_per_second.get_or_insert(drag);
        }
        let drag = if in_water {
            shell.water_drag_per_second.unwrap()
        } else {
            shell.drag_per_second.unwrap_or(0.0)
        };
        let mut flight = ballistic_step(from, shell.velocity, horizon, drag);
        let mut crosses_sea = false;
        let ends_in_water = flight.0[1] <= surface_at(flight.0[0], flight.0[2]);
        if in_water != ends_in_water && (in_water || from[1] > surface_at(from[0], from[2])) {
            let (mut low, mut high) = (0.0, horizon);
            for _ in 0..40 {
                let mid = (low + high) / 2.0;
                let point = ballistic_step(from, shell.velocity, mid, drag).0;
                if (point[1] <= surface_at(point[0], point[2])) == in_water {
                    low = mid;
                } else {
                    high = mid;
                }
            }
            let crossing = ballistic_step(from, shell.velocity, high, drag);
            if !inside_hull(crossing.0, actors, &shell.owner_id) {
                horizon = high;
                flight = crossing;
                crosses_sea = true;
            }
        }
        let end = flight.0;
        let mut nearest: Option<(usize, ShipContact)> = None;
        for (i, actor) in actors.iter().enumerate() {
            if actor.motion.id == shell.owner_id {
                continue;
            }
            let p = [actor.motion.x, actor.motion.y, actor.motion.z];
            let radius = actor.compiled.shell_radius;
            if !(0..3).all(|i| {
                from[i].min(end[i]) <= p[i] + radius && from[i].max(end[i]) >= p[i] - radius
            }) {
                continue;
            }
            if !segment_overlaps_box(
                world_to_local(from, actor.motion.pose()),
                world_to_local(end, actor.motion.pose()),
                actor.compiled.shell_center,
                actor.compiled.shell_size,
            ) {
                continue;
            }
            if let Some(hit) = ship_contacts(
                shell,
                from,
                end,
                actor,
                actor.definition(),
                &actor.compiled.contacts,
            )
            .into_iter()
            .next()
                && nearest.as_ref().is_none_or(|(_, h)| hit.t < h.t)
            {
                nearest = Some((i, hit));
            }
        }
        if let Some((t, point)) = first_land_hit(islands, fields, from, end)
            && nearest.as_ref().is_none_or(|(_, h)| t < h.t)
        {
            shell.position = point;
            events.push(DamageEvent {
                kind: "stopped".into(),
                position: point,
                message: "Shell struck the coast".into(),
                normal: Some([0.0, 1.0, 0.0]),
                shell: Some(ShellEffect::from_shell(shell)),
                ..Default::default()
            });
            return (Some(ProjectileEnd::Stopped), events);
        }
        if let Some((i, hit)) = nearest {
            let elapsed = if drag > 1e-9 {
                -(-drag * hit.t * travel_factor(horizon, drag)).ln_1p() / drag
            } else {
                horizon * hit.t
            };
            let at_hit = ballistic_step(from, shell.velocity, elapsed, drag);
            shell.penetration_mm = velocity_penetration(
                shell.penetration_mm,
                length(shell.velocity),
                length(at_hit.1),
            );
            shell.velocity = at_hit.1;
            shell.age += elapsed;
            remaining -= elapsed;
            let actor = &mut actors[i];
            let compiled = actor.compiled.clone();
            shell.position = local_to_world(hit.point, actor.motion.pose());
            let (stopped, event) =
                resolve_ship_contact(shell, &hit, actor, &compiled.definition, None);
            events.push(event);
            if stopped && shell.lodged.is_none() && shell.detonate_at_age.is_none() {
                return (Some(ProjectileEnd::Stopped), events);
            }
            continue;
        }
        shell.penetration_mm = velocity_penetration(
            shell.penetration_mm,
            length(shell.velocity),
            length(flight.1),
        );
        shell.velocity = flight.1;
        shell.position = end;
        shell.age += horizon;
        remaining -= horizon;
        if crosses_sea && in_water {
            shell.position[1] = surface_at(end[0], end[2]) + 1e-7;
            shell.water_drag_per_second = None;
        } else if crosses_sea {
            shell.position[1] = surface_at(end[0], end[2]) - 1e-7;
            events.push(DamageEvent {
                kind: "splash".into(),
                position: [end[0], surface_at(end[0], end[2]), end[2]],
                message: "Shell entering water".into(),
                shell: Some(ShellEffect::from_shell(shell)),
                ..Default::default()
            });
            if shell.he.is_some() {
                shell.detonate_at_age = Some(shell.age);
                continue;
            }
            let before = length(shell.velocity);
            shell.velocity = scale(shell.velocity, 0.75);
            shell.penetration_mm =
                velocity_penetration(shell.penetration_mm, before, length(shell.velocity));
            shell.water_drag_per_second = Some(
                1.0_f64.max(length(shell.velocity) * 0.04 * 0.38 / shell.caliber_m.max(0.007)),
            );
            if let Some(ap) = &shell.ap
                && shell.detonate_at_age.is_none()
                && shell.velocity[1].abs() / length(shell.velocity) > 0.15
            {
                shell.detonate_at_age = Some(shell.age + ap.fuze_delay_seconds);
            }
        }
    }
    (Some(ProjectileEnd::Expired), events)
}
